File: Canonical/227.basic-calculator-ii.go
package leetcode

// Calculate227 scans numbers and operators once. Addition/subtraction push signed
// values; multiplication/division immediately combine with the previous stack
// value to enforce precedence. Go integer division truncates toward zero.
//
// Time: O(n). Space: O(n).
func Calculate227(s string) int {
	stack := []int{}
	num := 0
	op := byte('+')
	for i := 0; i <= len(s); i++ {
		ch := byte('+')
		if i < len(s) {
			ch = s[i]
		}
		if ch == ' ' {
			continue
		}
		if ch >= '0' && ch <= '9' {
			num = num*10 + int(ch-'0')
			continue
		}
		if op == '+' {
			stack = append(stack, num)
		} else if op == '-' {
			stack = append(stack, -num)
		} else if op == '*' {
			stack[len(stack)-1] *= num
		} else {
			stack[len(stack)-1] /= num
		}
		op = ch
		num = 0
	}
	sum := 0
	for _, v := range stack {
		sum += v
	}
	return sum
}

File: companies/Amazon/227.basic-calculator-ii.py
"""
Approach: Single pass with a stack of signed terms.
Data structure: the stack stores additive terms after immediately resolving multiplication and division precedence.
Interview logic: when an operator is reached, apply the previous operator to the completed number. Addition/subtraction push signed numbers; multiplication/division update the previous stack term.
Complexity: O(n) time, O(n) space.
Tests and edge cases: spaces are ignored; integer division truncates toward zero; the final number is processed by appending a sentinel operator.
"""
from __future__ import annotations

# @lc code=start
class Solution:
    def calculate(self, s: str) -> int:
        stack = []
        num = 0
        op = '+'
        for ch in s + '+':
            if ch == ' ':
                continue
            if ch.isdigit():
                num = num * 10 + int(ch)
                continue
            if op == '+':
                stack.append(num)
            elif op == '-':
                stack.append(-num)
            elif op == '*':
                stack[-1] *= num
            else:
                prev = stack.pop()
                sign = -1 if prev < 0 else 1
                stack.append(sign * (abs(prev) // num))
            op = ch
            num = 0
        return sum(stack)
# @lc code=end
